playbooks/allocate_pubcloud.yml

```diff
@@ -4,40 +4,61 @@
   gather_facts: False
   vars:
     count: 1
+    keyname: "jenkins"
+    # retries are used as a mechanism for waiting for capacity
+    # as we don't have a mechanism for queuing jobs once
+    # they get this far
+    instance_boot_retries: 3
     inventory_path: "{{ lookup('env', 'WORKSPACE') }}/rpc-gating/playbooks/inventory"
     user_data_path: "{{ lookup('env', 'WORKSPACE') }}/rpc-gating/scripts/user_data_pubcloud.sh"
   tasks:
-    - name: Provision a set of public cloud instances
-      local_action:
-          module: rax
-          name: "{{ instance_name }}"
-          flavor: "{{ flavor }}"
-          image: "{{ image }}"
-          count: "{{ count }}"
-          key_name: "jenkins"
-          region: "{{ region }}"
-          # TODO(odyssey4me):
-          # In Ansible<2.3 there is a bug which prevents the rax module from
-          # reading user_data files. To work around that bug, we supply the
-          # contents of the user_data file rather than the path to it. This
-          # bug is fixed in 2.3, so once we upgrade to 2.3+, we can set
-          # user_data: user_data_path
-          user_data: "{{ lookup('file', user_data_path) }}"
-          meta:
-            build_config: core
-          wait: yes
-          wait_timeout: 900
-          auto_increment: no
-      register: rax
-      until: rax|success
-      retries: 5
-      delay: 60
 
-    - name: Display rax output on failure
-      debug:
-        var: rax
+    # Region should be a CSV string.
+    # Primary regions are those specified in the region var,
+    # they are tried first in a random order. After that,
+    # fallback regions are tried in a random order.
+
+    - name: Create primary region list
+      set_fact:
+        regions: "{{ region.split(',')|map('trim')|map('upper')|shuffle * instance_boot_retries }}"
```

Comment:
You should remove `""` from the list otherwise there are a few situations that will cause that to be one of the regions tried e.g. when `region` ends with a `,`.

Reply:
Done

Reply:
Added select() to reject empty strings.